ols/user_data_collection/data_collector.py

```diff
@@ -0,0 +1,313 @@
+        offline_token: Offline token from the Customer Portal.
+
+    Returns:
+        Refresh token.
+    """
+    if INGRESS_ENV == "stage":
+        url = "https://sso.stage.redhat.com/"
+    else:
+        url = "https://sso.redhat.com/"
+    endpoint = "auth/realms/redhat-external/protocol/openid-connect/token"
+    data = {
+        "grant_type": "refresh_token",
+        "client_id": "rhsm-api",
+        "refresh_token": offline_token,
+    }
+
+    response = requests.post(url + endpoint, data=data, timeout=REDHAT_SSO_TIMEOUT)
+    if response.status_code == requests.codes.ok:
+        return response.json()["access_token"]
+    else:
+        raise Exception(f"Failed to generate access token. Response: {response.json()}")
+
+
+def get_cloud_openshift_pull_secret() -> str:
+    """Get the pull secret token from the cluster."""
+    kubernetes.config.load_incluster_config()
+    v1 = kubernetes.client.CoreV1Api()
+
+    # get the pull-secret from openshift-config namespace
+    secret = v1.read_namespaced_secret("pull-secret", "openshift-config")
+    dockerconfigjson = secret.data[".dockerconfigjson"]
+    dockerconfig = json.loads(base64.b64decode(dockerconfigjson).decode("utf-8"))
+
+    return dockerconfig["auths"]["cloud.openshift.com"]["auth"]
+
+
+def get_cluster_id() -> str:
+    """Get the cluster_id from the cluster."""
+    kubernetes.config.load_incluster_config()
+    custom_objects_api = kubernetes.client.CustomObjectsApi()
+
+    # get the cluster version object
+    version_data = custom_objects_api.get_cluster_custom_object(
+        "config.openshift.io", "v1", "clusterversions", "version"
+    )
+
+    return version_data["spec"]["clusterID"]
+
+
+def collect_ols_data_from(location: str) -> list[pathlib.Path]:
+    """Collect files from a given location.
+
+    Args:
+        location: Path to the directory to be searched for files.
+
+    Returns:
+        List of paths to the collected files.
+
+    Only JSON files from the 'feedback' and 'transcripts' directories are collected.
+    """
+    files = []
+
+    files += list(pathlib.Path(location).glob("feedback/*.json"))
+    files += list(pathlib.Path(location).glob("transcripts/*/*/*.json"))
+
+    return files
+
+
+def package_files_into_tarball(
+    file_paths: list[pathlib.Path], path_to_strip: str
+) -> io.BytesIO:
+    """Package specified directory into a tarball.
+
+    Args:
+        file_paths: List of paths to the files to be packaged.
+        path_to_strip: Path to be stripped from the file paths (not
+            included in the archive).
+
+    Returns:
+        BytesIO object representing the tarball.
+    """
+    tarball_io = io.BytesIO()
+
+    with tarfile.open(fileobj=tarball_io, mode="w:gz") as tar:
+        # arcname parameter is set to a stripped path to avoid including
+        # the full path of the root dir
+        for file_path in file_paths:
+            tar.add(file_path, arcname=file_path.as_posix().replace(path_to_strip, ""))
```

Comment:
is there a way to ensure we do not add or follow symlinks?  we should do so as symlinks are one way to exploit a system into gathering files we didn't intend.